opentelemetry/src/sdk/trace/provider.rs

```diff
@@ -139,10 +142,18 @@ impl Builder {
 
     /// Create a new provider from this configuration.
     pub fn build(self) -> TracerProvider {
+        let mut config = self.config;
+        let sdk_provided_resource = Resource::from_detectors(
+            Duration::from_secs(0),
+            vec![Box::new(SdkProvidedResourceDetector)],
+        );
+        if config.resource.is_none() {
+            config.resource = Some(Arc::new(sdk_provided_resource))
```

Comment:
Sounds reasonable. But I'd argue this usually only happens when the application starts so it may not impact performance that much. I can create `sdk_provided_resource` only when there isn't a resource already configured so that could reduce some overhead.

Reply:
Was thinking more about the clone each time a span is exported, could check https://github.com/open-telemetry/opentelemetry-rust/blob/efbc842c11ab416309032ae794c09f6edd851be7/opentelemetry/benches/trace.rs#L27 at some point

Reply:
Yeah, you are right. This will add an attribute to each span by default. But users can avoid that by passing an empty resource in the config.

Reply:
I ran the bench. With default resource & required attributes:
```
start-end-span/always-sample                                                                             
                        time:   [536.05 ns 538.61 ns 541.21 ns]
                        change: [+18.240% +18.917% +19.533%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
start-end-span/never-sample                                                                            
                        time:   [166.50 ns 166.98 ns 167.66 ns]
                        change: [+0.1332% +0.4798% +0.8373%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe


```

Reply:
Hm I wonder if there are other alternatives here, maybe a `&'static Resource` to avoid all the `Arc::clone`'s for data that never changes?

Reply:
Hmm, I don't think that works for resources created from environment variables? Because we cannot know the value in the compile time. I am thinking if users can somehow explicitly says they don't want any resource maybe we could avoid the overhead.

Reply:
@jtescher I guess a workaround here is to explicitly pass an empty resource into the config to indicate users don't want any resource including the one provided by SDK by default. Then we won't have any performance impact on users.

Reply:
Hm yeah that's a possibility. Wonder if there should be an explicit method in the builder or otherwise more discoverable method. People doing comparisons like https://github.com/tikv/minitrace-rust/blob/master/benches/compare.rs will likely miss the possible optimizations otherwise.

Reply:
Added a `with_no_resource` method for that purpose.

Reply:
ok can merge this for now and add a check in the 1.0 readiness to see if breaking changes are necessary for perf reasons